iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -17,45 +17,149 @@
 
 #include "iceoryx_posh/internal/roudi/service_registry.hpp"
 
+#include <algorithm>
+
 namespace iox
 {
 namespace roudi
 {
-void ServiceRegistry::add(const capro::IdString_t& service, const capro::IdString_t& instance)
+cxx::expected<ServiceRegistry::Error> ServiceRegistry::add(const capro::ServiceDescription& serviceDescription) noexcept
 {
-    cxx::set::add(m_serviceMap[service].instanceSet, instance);
+    if (!serviceDescription.isValid())
+    {
+        return cxx::error<Error>(Error::SERVICE_DESCRIPTION_INVALID);
+    }
+
+    // Forbid duplicate service descriptions entries
+    for (auto& element : m_serviceDescriptionVector)
+    {
+        if (element == serviceDescription)
+        {
+            return cxx::error<Error>(Error::SERVICE_DESCRIPTION_ALREADY_ADDED);
+        }
+    }
+
+    if (!m_serviceDescriptionVector.push_back(serviceDescription))
+    {
+        return cxx::error<Error>(Error::SERVICE_REGISTRY_FULL);
+    }
+    m_serviceMap.insert({serviceDescription.getServiceIDString(), m_serviceDescriptionVector.size() - 1});
+    m_instanceMap.insert({serviceDescription.getInstanceIDString(), m_serviceDescriptionVector.size() - 1});
+    return cxx::success<>();
 }
 
-void ServiceRegistry::remove(const capro::IdString_t& service, const capro::IdString_t& instance)
+bool ServiceRegistry::remove(const capro::ServiceDescription& serviceDescription) noexcept
 {
-    cxx::set::remove(m_serviceMap[service].instanceSet, instance);
+    if (!serviceDescription.isValid())
+    {
+        return false;
+    }
+
+    bool removedElement{false};
+    uint64_t index = 0U;
+    for (auto iterator = m_serviceDescriptionVector.begin(); iterator != m_serviceDescriptionVector.end();)
+    {
+        if (m_serviceDescriptionVector[index] == serviceDescription)
+        {
+            m_serviceDescriptionVector.erase(iterator);
+            removedElement = true;
+            // There can be not more than one element
+            break;
+        }
+        index++;
+        iterator++;
+    }
+
+    auto removeIndexFromMap = [](std::multimap<capro::IdString_t, uint64_t>& map, uint64_t index) {
+        for (auto it = map.begin(); it != map.end();)
```

Comment:
Those kind of iteration defeats the purpose of having efficient search  structures and is not necessary here.
If we want to remove a `Servicedescription` with `service` and `instance` we do the following (I prefer pseudo code to explain, it is easy to write this in C++).
1. find indices corresponding to service in `m_serviceMap, call those I1
2. find indices corresponding to instance in `m_instanceMap`, call those I2
3. compute the intersection I = I1 \intersection I2. This is one (sevice exists) or no element
4. If I is empty, the service does not exist and we have nothing to remove and can return.
5. Otherwise we remove the single element from the container and the maps.

Removing from the maps is rather tricky to do efficiently right now, but the prefix tree will support an operation that allows to remove a specific value to be removed from a specific key.

This is needed since e.g. for the `m_serviceMap` we already know the key, service. And after the intersection, we also know the value, the only value in the intsersection I. `std::map` does not efficiently support this though, so for the time being we could iterate over the map.

But for the container (currently the `m_serviceDescriptionVector`) the intersection approach must be used for efficiency. Otherwise the operations all degenerate into linear searches.

Note that this approach naturally generalizes to an arbitrary number of search attributes.

Also note that this is partly the algorithm for finding as well, at the internal level where we still operate with indices in our container (which the client will of course not see). So this can be suitably factored out. After all, removing is finding the indices to be removed and removing corresponding data. Finding for the client is just finding the indices and translating them to the actual data to be returned.

Reply:
Ok, got your point.

Unfortunately with the `cxx::vector` that is not possible as `erase()` expects an iterator currently. So I have to iterate over the complete vector and can't use two `equal_range` searches with intersection which I used in `ServiceRegistry::find()`. Once we have an ready-made alternative like the ObjectPool I'm happy to refactor things, but for the very first version I'd like to stick with the `cxx::vector`. Let's optimise this when integrating #829 into the `ServiceRegistry`. I added a todo in #415.